Hourly Compaction

zliu41 opened the pull request:
How this works:

- There is a `FolderState` associated with each input folder. Possible values are `YES`, `NO`, `PENDING`, `COMPLETE`. Initially the status of all input folders is `NO`.
- When compaction starts, it submits data completeness verification requests in a thread pool, one request per input folder, in the order of their priorities (high priority = 3.0, normal = 2.0, low = 1.0). After data completeness verification for a folder starts, the folder status changes to `PENDING`.
- Meanwhile, a job submitter submits compaction MR jobs for the input folders to another thread pool. The compaction MR job of an input folder is only submitted if the folder status is `PENDING` or `YES`, and there is not already a compaction MR job running for this input folder. If the status is `NO`, there's no point to run the compaction MR job since we don't know if the data is complete or not. If the status is `COMPLETE`, it means compaction for this input folder has finished (which may have either succeeded or failed), and the job submitter will no longer submit MR jobs for this input folder. If there's already a compaction MR job running for this input folder, the job submitter will not submit it again until the current run finishes.
 - If, when a compaction finishes, the status of the input folder is still `PENDING`, it will not publish the data and will discard the work it has done.
- Once the data completeness verification for a folder finishes:
 - if the data is complete, the status of the folder is changed to `YES`, and the thread which is running the compaction job for this folder (if exists) is notified, which will proceed to publish the data. 
 - if the data is incomplete, the status of the folder is changed to `NO`, and the thread which is running the compaction job for this folder will be asked to terminate asap without publishing the data. Another data completeness verification for this folder will be submitted again with priority reduced by 1/3.
- Once the compaction MR job for a folder finishes, if the job successfully published the data, or failed for reasons other than data incompleteness, the status of the input folder will be changed to COMPLETE, and the job submitter will no longer submit jobs for this input folder. Otherwise, the job submitter will attempt to re-submit the compaction MR job for this input folder with priority reduced by 1/3 (it will not re-submit the job until the input folder status is changed to PENDING or YES, as mentioned before).

zliu41 commented:
@liyinan926 @chavdar Could you please review?

chavdar commented on gobblin-compaction/src/main/java/gobblin/compaction/Compactor.java:
```diff
@@ -26,5 +26,5 @@
    * @throws IOException
    */
-  public void compact() throws IOException;
+  public void compact() throws Exception;
```
Please don't use Exception in the throws clause as it defeats the purpose of static exception checking. List the type of exceptions that can happent.

chavdar commented on gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactor.java:
```diff
@@ -46,16 +50,46 @@
+    // An input folder with status NO indicates that the verification of its data completeness has not started.
+    // If data completeness verification is enabled, compaction should not start on an input folder if its status is
+    // NO, since we won't know whether its data is complete or not.
+    NO,
```
Please rename to something more meaningful like VERIFICATION_PENDING.

liyinan926 replied:
I would further suggest creating two enums, one named `VerificationState` and the other named `CompactionState`.

chavdar commented on gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactor.java:
```diff
@@ -46,16 +50,46 @@
+    // Compaction may start on a folder with PENDING status. If, when the compaction job finishes, its status is still
+    // PENDING, the job won't publish data. If during the compaction job, the data verification finishes and the status
+    // changes to NO, the compaction job will be killed.
+    PENDING,
```
How about VERIFICATION_STARTED

chavdar commented on gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactor.java:
```diff
@@ -46,16 +50,46 @@
+    PENDING,
+
+    // YES indicates that data of an input folder is complete.
+    YES,
```
How about VERIFICATION_PASSED

zliu41 replied:
I liked YES and NO from aesthetical point of view but it's fine if you prefer VERIFICATION_PASSED

chavdar replied:
IMO, the problem with YES and NO  is that you can't really know what they mean unless you read the docs. I think it is better when the code is self-explanatory.

chavdar commented on gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactor.java:
```diff
@@ -46,16 +50,46 @@
+    YES,
+
+    // COMPLETE indicates that compaction of an input folder has finished, which may have either succeeded or failed.
+    COMPLETE
```
How about COMPACTION_COMPLETE

chavdar commented on gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/AbstractDataCompletenessVerifier.java:
```diff
@@ -0,0 +1,72 @@
+public abstract class AbstractDataCompletenessVerifier
+    implements Runnable, Comparable<AbstractDataCompletenessVerifier> {
+
+  protected final MRCompactor mrCompactor;
```
I don't think we have the right abstractions here. The verifier should not be modifying the state of the compactor directly. I suggest that you define a callback interface which can have the same methods. MRCompactor  can pass an implementation of that interface which will modify its state.

liyinan926 replied:
You can also take a look at Guava's `EventBus` class and possible use that as the channel for notification between a verifier and the compactor.

liyinan926 commented on gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/AbstractDataCompletenessVerifier.java:
```diff
@@ -0,0 +1,72 @@
+  protected final State jobProps;
+  protected final double priority;
+
+  public AbstractDataCompletenessVerifier(MRCompactor compactor, String inputFolder, State jobProps, Double priority) {
```
`Double` => `double`.

chavdar replied:
I don't think the verifier object should be on a per-folder/dataset level. This precludes the ability to pool multiple completeness checks to the audit backend.

I also think we should decouple the completeness read from the check if the completeness value meets the publish threshold. The former may also be useful for other purposes like monitoring.

I suggest you define a DatasetAuditCountReader interface or something along that lines. It should define a method like ListenableFuture<Long> readAuditCounts(dataset). ListenableFuture allows to add a callback on success or failure which can then modify the state of MRCompactor. Whether there is a threadpool to run the audit count requests there is some kind of elevator algorithm or so should be completely up to the reader implementation. MRCompactor shouldn't be concerned with thread pools and so on.

You can also have a DatasetAuditCountCompletenessChecker which has a checkCompleteness(dataset, count) method to verify the completeness. This should be fairly light-weight and should probably be done synchronously. 

With this model, I don't think you need afterVerificationStarts() and  afterVerificationFinishes(). The process should be driven by MRCompactor and it could provide the necessary callbacks to drive the state transitions.

I also suggest that you add some abstraction around the input folder. I am not sure what this folder is. Is it an HDFS folder? What is the namenode, etc. I think the interfaces should use a dataset interface instead. It can have the attributes that we discussed for the dataset publication. Also consider the use case of running compaction on multiple folders.

zliu41 replied:
`AbstractDataCompletenessVerifier` doesn't contradict with decoupling read and check. It's just an abstract class; its subclasses can first read from kafka audit, then check. Those implementations will be in an RB since they are internal.

zliu41 replied:
`Double` is used instead of `double` since `getDeclaredConstructors()` doesn't support primitive types.

chavdar replied:
To me AbstractDataCompletenessVerifier does not define a meaningful API for classes to override/implement. There are just some conventions which are very much tied to the MRCompactor implementation. We need an interface independent of MRCompactor.

chavdar commented on gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/AbstractDataCompletenessVerifier.java:
```diff
@@ -0,0 +1,72 @@
+  public AbstractDataCompletenessVerifier(MRCompactor compactor, String inputFolder, State jobProps, Double priority) {
+    this.mrCompactor = compactor;
+    this.inputFolder = inputFolder;
+    this.jobProps = jobProps;
```
Why is that needed? I don't see it being used any where.

zliu41 replied:
Some implementations of data completeness verifier may want to use properties in `jobProps` in verifying completeness..

chavdar replied:
Then the implementations that may need it should just declare it.

liyinan926 commented on gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/AbstractDataCompletenessVerifier.java:
```diff
@@ -0,0 +1,72 @@
+  }
+
+  /**
+   * This method should be called by subclasses right after data completeness verification starts for the input folder.
```
I would suggest implementing the `run` like the following so this and the method below are guaranteed to be called. The actual logic of verification should be implemented by the subclasses in the method `verify`.

```
public void run() {
  afterVerificationStarts();
  boolean dataComplete = verify();
  afterVerificationFinishes(dataComplete);    
}

protected abstract boolean verify();
```

zliu41 replied:
`afterVerificationStarts();` should be called after the `verify()` starts, not before.

chavdar commented on gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactor.java:
```diff
@@ -65,8 +99,14 @@
   private final FileSystem fs;
-  private final ExecutorService executorService;
-  private final List<Future<?>> futures;
+  private final Map<String, State> inputFolderJobProps;
```
Please define a record type instead of having three different maps with the same key.

chavdar commented on gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactor.java:
```diff
@@ -78,8 +118,18 @@ public MRCompactor(Properties props) throws IOException {
+    this.inputFolderStates = Maps.newConcurrentMap();
+    this.foldersBeingProcessed = Sets.newConcurrentHashSet();
+    this.jobExecutor = createJobExecutor();
+    if (shouldVerifyDataCompleteness()) {
```
We should allow controlling the completeness checks on a per-dataset level.

chavdar commented on gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactor.java:
```diff
@@ -278,40 +319,286 @@ MRCompactorJobPropCreator getJobPropCreator(String topic, Path topicInputDir, Pa
+  }
+
+  private boolean shouldVerifyDataCompleteness() {
+    return this.state.getPropAsBoolean(ConfigurationKeys.COMPACTION_SHOULD_VERIFY_DATA_COMPLETENESS,
```
Please don't check at the configuration properties map in methods. Check only in the constructor and initialize the necessary member variables. Otherwise, figuring out all configuration that is used by a class is a major PITA. That is especially true when the configs have default values.

chavdar commented on gobblin-api/src/main/java/gobblin/configuration/ConfigurationKeys.java:
```diff
@@ -475,6 +475,15 @@
   public static final String COMPACTION_OVERWRITE_OUTPUT_DIR = COMPACTION_PREFIX + "overwrite.output.dir";
   public static final boolean DEFAULT_COMPACTION_OVERWRITE_OUTPUT_DIR = false;
   public static final String COMPACTION_JARS = COMPACTION_PREFIX + "jars";
+  public static final String COMPACTION_DATA_COMPLETENESS_VERIFICATION_THREAD_POOL_SIZE =
```
These configs are specific to the MRCompactor. Please declare them in the class that uses them.

liyinan926 commented on gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactor.java:
```diff
@@ -278,40 +319,286 @@ MRCompactorJobPropCreator getJobPropCreator(String topic, Path topicInputDir, Pa
+          MRCompactor.this.compactionJobThrowables.put(jobRunner.getInputPath().toString(), t);
+        }
+      } else {
+        LOG.error(String.format("Runnable expected to be instance of %s, actual %s",
```
You probably want to use a `Preconditions.checkArgument()` to make sure the right type is seen here.

liyinan926 commented on gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactor.java:
```diff
@@ -278,40 +319,286 @@ MRCompactorJobPropCreator getJobPropCreator(String topic, Path topicInputDir, Pa
+    Stopwatch sw = Stopwatch.createStarted();
+
+    while (!MRCompactor.this.foldersBeingProcessed.isEmpty()) {
+      for (Iterator<String> it = MRCompactor.this.foldersBeingProcessed.iterator(); it.hasNext();) {
```
You can use a `while` loop here that will look more natural.

liyinan926 commented on gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactor.java:
```diff
@@ -278,40 +319,286 @@ MRCompactorJobPropCreator getJobPropCreator(String topic, Path topicInputDir, Pa
+
+  private void shutdownExecutors() {
+    LOG.info("Shutting down Executors");
+    this.jobExecutor.shutdownNow();
```
You probably can use `ExecutorUtils.shutdownExecutorService()`.

liyinan926 commented on gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactorJobRunner.java:
```diff
@@ -262,4 +308,33 @@ private void moveTmpPathToOutputPath() throws IOException {
+  /**
+   * Tell the MRCompactorJobRunner that it can go ahead and publish the data.
+   */
+  public void go() {
```
Will it possible for more than one threads to call `go` or `stop` below?

zliu41 replied:
No, only one thread verifies data completeness for a folder.

liyinan926 commented on gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactor.java:
```diff
@@ -46,16 +50,46 @@
+  private static final double LOW_PRIORITY = 1.0;
+  private static final double PRIORITY_REDUCTION_FACTOR = 1.0 / 3.0;
+  private static final long COMPACTION_JOB_WAIT_INTERVAL_SECONDS = 10;
+  private static final Map<String, Job> RUNNING_MR_JOBS = Maps.newConcurrentMap();
```
This probably should just be `final` instead of `static final`.

zliu41 replied:
Then I'll need to pass an instance of MRCompactor to `MRCompactorJobRunner` which creates the Hadoop job.

liyinan926 replied:
Then can we move this `Map` to `MRCompactorJobRunner` which seems a better place to host it?

liyinan926 commented on gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactor.java:
```diff
@@ -125,30 +176,23 @@ private int getThreadPoolSize() {
+        shutdownExecutors();
+      } finally {
         deleteDependencyJars();
-        for (Future<?> future : this.futures) {
```
Instead of storing `Throwable`s in `compactionJobThrowables` and check them in `checkForThrowables`, you can probably use `ListeningFuture` and `ListenableFuture` and the callback mechanism around it. Take a look at https://github.com/google/guava/wiki/ListenableFutureExplained.

zliu41 replied:
`Future` is no longer used. I need to use `ThreadPoolExecutor.execute()` which does not return `Future`, since `submit()` cannot prioritize the tasks using a priority queue.

zliu41 commented:
@chavdar @liyinan926 updated (cbc38b6).
---Summary---

Big changes:
- Added `gobblin.compaction.Dataset` class. This is a temporary solution; after Gobblin dataset is merged, I'll put it into the class hierarchy.
- In addition to verifying data completeness using `DataCompletenessVerifier` (via Kafka audit), the `MRCompactorJobRunner` also performs a self-verification: when compaction of a dataset completes, if it finds new files that were added to the input folder after compaction started, it will discard the compacted data and re-do the compaction.
- The compaction MR job of a dataset is no longer required to start after the data completeness verification of the dataset starts. Now it submits compaction MR jobs to one thread pool, and the completeness verification jobs to another thread pool, with no particular order between them.
- `MRCompactor` no longer manages the thread pool for data completeness verification. That thread pool is now managed by `DataCompletenessVerifier`.
- `DataCompletenessVerifier` no longer holds a reference to `MRCompactor`. Instead, it returns a `ListenableFuture<Results>` object, which `MRCompactor` can use to add callbacks and modify its states. This also means that data verification is no longer performed in the order of dataset priorities, since ExecutorService.submit() doesn't support prioritization through a priority queue (compaction MR jobs are stilled ordered by dataset priorities).
- Added a `SimulatedDataCompletenessVerifier` class which randomly makes decisions.

Small changes:
- Now it is possible to verify the completeness of multiple datasets together by using `DataCompletenessVerifier.verify(Iterable<Dataset>)`. This enables multiple datasets to share a single SQL query against Kafka audit.
- Now it is possible to specify which datasets need completeness verification by using a whitelist and a blacklist.
- Now one can specify the timeout for verifying data completeness, e.g. 30min. If after 30 minutes from the time the compaction starts, there are still some datasets whose completeness are not verified, those datasets will be given up, and their running verification jobs (if exists) will be killed. For these datasets, it is configurable whether to not compact them, or compact them anyway without completeness guarantee.
- Addressed other comments in the review.

zliu41 commented:
@liyinan926 updated 7edb31a

chavdar commented on gobblin-compaction/src/main/java/gobblin/compaction/verify/DataCompletenessVerifier.java:
```diff
@@ -0,0 +1,163 @@
+  }
+
+  private int getDataCompletenessVerificationThreadPoolSize() {
+    return this.props.getPropAsInt(ConfigurationKeys.COMPACTION_DATA_COMPLETENESS_VERIFICATION_THREAD_POOL_SIZE,
```
This should be a member variable in the constructor. Please avoid accessing the configuration object outside the constructor. It makes it very difficult to figure out all configuration properties. It is also error prone if there are default values involved.

chavdar commented on gobblin-api/src/main/java/gobblin/configuration/ConfigurationKeys.java:
```diff
@@ -470,20 +469,32 @@
   public static final boolean DEFAULT_COMPACTION_OVERWRITE_OUTPUT_DIR = false;
-  public static final String COMPACTION_JARS = COMPACTION_PREFIX + "jars";
-  public static final String COMPACTION_TRACKING_EVENTS_NAMESPACE = COMPACTION_PREFIX + "tracking.events";
+  public static final String COMPACTION_DATA_COMPLETENESS_VERIFICATION_THREAD_POOL_SIZE =
```
Please bundle configuration properties with the code that uses them rather than dumping them in ConfigurationKeys .

zliu41 replied:
All config properties that can be set by the user are in ConfigurationKeys. All internal properties have been moved to MRCompactor.

chavdar replied:
That's the part I don't get. What does it buy us? We have so many of these options (363 in my version)  that I doubt anyone will say "Let me see what knobs and levers are out there". More likely, the question is "I need to figure out how to configure component X". Given that the property naming is not very consistent, I think the starting point is always going to be the implementation. I think ConfigurationKeys should be only for keys that are shared across components. This is most likely state properties and not real configuration properties.

I also suggest at renaming the configuration properties so that they share a more specific prefix than compaction, e.g. compaction.completeness_verification or similar. Then, "compaction.should.verify.data.completeness" can become "compaction.completeness_verification.enabled", "compaction.data.completeness.verification.class" can become "compaction.completeness_verification.verificatier_class" and so on. I find the current "English phrase with dots" naming scheme confusing as the same concepts can be expressed in different ways.

This also reminds me of another issue -- please add javadocs for all configuration properties what they mean. Looking just at the configs I don't think I can guess what "compaction.num.datasets.verified.together" does.

chavdar commented on gobblin-azkaban/src/main/java/gobblin/azkaban/AzkabanCompactionJobLauncher.java:
```diff
@@ -42,15 +44,22 @@ public AzkabanCompactionJobLauncher(String jobId, Properties props) {
-    String compactorClassName =
-        this.properties.getProperty(ConfigurationKeys.COMPACTION_COMPACTOR_CLASS,
-            ConfigurationKeys.DEFAULT_COMPACTION_COMPACTOR_CLASS);
+    String compactorClassName = this.properties.getProperty(ConfigurationKeys.COMPACTION_COMPACTOR_CLASS,
```
This logic should be in the constructor.

chavdar commented on gobblin-azkaban/src/main/java/gobblin/azkaban/AzkabanCompactionJobLauncher.java:
```diff
@@ -42,15 +44,22 @@ public AzkabanCompactionJobLauncher(String jobId, Properties props) {
   public void run() throws Exception {
     Class<? extends Compactor> compactorClass = getCompactorClass();
-    compactorClass.getDeclaredConstructor(Properties.class).newInstance(this.properties).compact();
+    this.compactor = compactorClass.getDeclaredConstructor(Properties.class).newInstance(this.properties);
```
I'd move this to the constructor and make it final. Constructor should not leave the object in semi-initialized state.

chavdar commented on gobblin-compaction/src/main/java/gobblin/compaction/verify/SimulatedDataCompletenessVerifier.java:
```diff
@@ -0,0 +1,96 @@
+
+/**
+ * A class that simulates data completeness verification for Gobblin compaction. It randomly decides
+ * whether the verification of a {@link Dataset} passes or fails.
```
What is the purpose of the class? Is it for testing? If yes, In general, having test classes rely on random behavior is not ideal as they make reproducing errors hard. At the very least, there should be enough audit trail about all the random decisions made.

chavdar commented on gobblin-compaction/src/main/java/gobblin/compaction/verify/SimulatedDataCompletenessVerifier.java:
```diff
@@ -0,0 +1,96 @@
+    try {
+      Thread.sleep(verifTime * 1000);
+    } catch (InterruptedException e) {
+      LOG.error("Interrupted", e);
```
Nit: if the wait was interrupted. It's probably best to propagate a runtime exception as something is being shutdown/stopped.

liyinan926 replied:
Should also do `Thread.currentThread.interrupt()` to restore the interruption flag.

chavdar commented on gobblin-compaction/src/main/java/gobblin/compaction/verify/SimulatedDataCompletenessVerifier.java:
```diff
@@ -0,0 +1,96 @@
+    Preconditions.checkState(minTime >= 0, SIMULATED_VERIFIER_MIN_VERIFICATION_TIME_SEC + " should be nonnegative.");
+    Preconditions.checkState(minTime <= maxTime, SIMULATED_VERIFIER_MIN_VERIFICATION_TIME_SEC
+        + " should be no more than " + SIMULATED_VERIFIER_MAX_VERIFICATION_TIME_SEC);
+    long verifTime = RANDOM.nextInt(maxTime - minTime) + minTime;
```
typo: s/verifTime/verifyTime/

chavdar commented on gobblin-compaction/src/main/java/gobblin/compaction/verify/DataCompletenessVerifier.java:
```diff
@@ -0,0 +1,163 @@
+
+  private ListeningExecutorService getExecutorService() {
+    int threadPoolSize = getDataCompletenessVerificationThreadPoolSize();
+    return MoreExecutors.listeningDecorator(Executors.newFixedThreadPool(threadPoolSize));
```
Using a fixed thread pools ties up resources (threads) when not needed. Have a look at gobblin.util.executors.ScalingThreadPoolExecutor . We should use a thread pool with sizes (0, threadPoolSize) so that the thread pool can grow and shrink as needed.

chavdar commented on gobblin-compaction/src/main/java/gobblin/compaction/verify/DataCompletenessVerifier.java:
```diff
@@ -0,0 +1,163 @@
+    Preconditions.checkArgument(this.props.contains(ConfigurationKeys.COMPACTION_DATA_COMPLETENESS_VERIFICATION_CLASS),
+        "Missing required property " + ConfigurationKeys.COMPACTION_DATA_COMPLETENESS_VERIFICATION_CLASS);
+    try {
+      @SuppressWarnings("unchecked")
```
The class object creation should happen in the constructor.

chavdar commented on gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactor.java:
```diff
@@ -297,39 +368,355 @@ MRCompactorJobPropCreator getJobPropCreator(String topic, Path topicInputDir, Pa
+    for (Dataset dataset : datasetsWithThrowables) {
+      numDatasetsWithThrowables++;
+      for (Throwable t : dataset.throwables()) {
+        LOG.error("Error processing dataset " + dataset, t);
```
We should emit tracking events.

zliu41 commented:
@chavdar updated a5a11a7

chavdar commented on gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactor.java:
```diff
@@ -294,39 +373,357 @@ MRCompactorJobPropCreator getJobPropCreator(String topic, Path topicInputDir, Pa
+      } else {
+
+        // Compaction job of a dataset has failed with a throwable.
+        jobRunner.getDataset().setState(COMPACTION_COMPLETE);
```
For these kind of state transitions where multiple attributes need to be changed, it is better to have a helperMethod, e.g. completeWithError(Throwable) to make sure that the object is not in an inconsistent state.

chavdar commented on gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactorJobRunner.java:
```diff
@@ -75,57 +98,73 @@
+
+  protected MRCompactorJobRunner(Dataset dataset, FileSystem fs, Double priority) {
+    this.dataset = dataset;
+    this.jobProps = dataset.jobProps();
```
This is probably going to change once you integrate with the generic dataset but I think it looks unusual that the dataset is coupled with the job properties.

zliu41 replied:
Yes this will change. I think it makes sense that a compaction dataset has a compaction job props, and it's better than adding another Map in `MrCompactor` from dataset to jobprops.

chavdar replied:
Ok. I'll wait to see for the new PR with the integrated datasets. In general, if those are associated with the dataset, they should be "datasetProps", not "jobProps". 

Further,  as an editorial comment, I think we have way too many property collections. It makes it hard to figure out what is the life-cycle of the collections and the properties, which are read/only and which are read/write, and all kinds of other data dependencies. I'd love to see more well-type java objects and interfaces and less property collections.  This will also make us think a little bit more what the API is instead of just dumping stuff into a collection.

chavdar commented on gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactor.java:
```diff
@@ -294,39 +373,357 @@ MRCompactorJobPropCreator getJobPropCreator(String topic, Path topicInputDir, Pa
+          Optional<MRCompactorJobRunner> jobRunner =
+              Optional.fromNullable(MRCompactor.this.jobRunnables.get(result.dataset()));
+
+          if (result.status() == DataCompletenessVerifier.Results.Result.Status.PASSED) {
```
I think it is better to have a switch statement here. If someone later adds a new state, there will be a warning that there is a case that has not been processed.

chavdar commented on gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactor.java:
```diff
@@ -294,39 +373,357 @@ MRCompactorJobPropCreator getJobPropCreator(String topic, Path topicInputDir, Pa
+      try {
+        Thread.sleep(TimeUnit.SECONDS.toMillis(COMPACTION_JOB_WAIT_INTERVAL_SECONDS));
+      } catch (InterruptedException e) {
+        LOG.error("Interruppted while waiting", e);
```
InterruptedExceptions should not be swallowed. Most likely someone is trying to shutdown the service. In that case the code should exit gracefully.

chavdar commented on gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactor.java:
```diff
@@ -294,39 +373,357 @@ MRCompactorJobPropCreator getJobPropCreator(String topic, Path topicInputDir, Pa
+      for (Throwable t : dataset.throwables()) {
+        LOG.error("Error processing dataset " + dataset, t);
+        CompactionSlaEventHelper.populateState(dataset.jobProps(), Optional.<Job> absent(), this.fs);
+        new SlaEventSubmitter(eventSubmitter, "CompactionFailed", dataset.jobProps().getProperties()).submit();
```
Actually, reviewing more in depth the rest of the code, I think we need the event emission to happen on certain dataset state transitions, like like VERIFIED, GIVEN_UP, etc.

zliu41 commented:
@chavdar updated 980251b

chavdar commented:
LGTM
